launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,517 @@
+        );
+
+        let confirmation_start_epoch = self.confirmation_period_start_epoch().get();
+        let confirmation_period_in_epochs = self.confirmation_period_in_epochs().get();
+        let confirmation_end_epoch = confirmation_start_epoch + confirmation_period_in_epochs;
+
+        require!(
+            current_epoch > confirmation_start_epoch,
+            "Confirmation period has not started yet"
+        );
+        require!(
+            current_epoch <= confirmation_end_epoch,
+            "Confirmation period has ended"
+        );
+
+        let ticket_payment_token = self.ticket_payment_token().get();
+        let ticket_price = self.ticket_price().get();
+        let total_ticket_price = Self::BigUint::from(nr_tickets_to_confirm) * ticket_price;
+
+        require!(
+            payment_token == ticket_payment_token,
+            "Wrong payment token used"
+        );
+        require!(payment_amount == total_ticket_price, "Wrong amount sent");
+
+        let (first_ticket_id, last_ticket_id) = self.try_get_ticket_range(&caller)?;
+        let nr_tickets = last_ticket_id - first_ticket_id + 1;
+
+        require!(
+            nr_tickets >= nr_tickets_to_confirm,
+            "Trying to confirm too many tickets"
+        );
+
+        let current_generation = self.current_generation().get();
+        let mut actual_confirmed_tickets = 0;
+
+        for ticket_id in first_ticket_id..=last_ticket_id {
+            let ticket_status = self.ticket_status().get(ticket_id);
+            if !ticket_status.is_winning(current_generation) {
+                continue;
+            }
+
+            self.set_ticket_status(
+                ticket_id,
+                TicketStatus::Confirmed {
+                    generation: current_generation,
+                },
+            );
+            actual_confirmed_tickets += 1;
+
+            if actual_confirmed_tickets == nr_tickets_to_confirm {
+                break;
+            }
+        }
+
+        require!(
+            actual_confirmed_tickets == nr_tickets_to_confirm,
+            "Couldn't confirm all tickets"
+        );
+
+        Ok(OperationCompletionStatus::Completed)
+    }
+
+    #[endpoint(claimLaunchpadTokens)]
+    fn claim_launchpad_tokens(&self) -> SCResult<OperationCompletionStatus> {
+        let caller = self.blockchain().get_caller();
+        require!(
+            !self.blacklist().contains(&caller),
+            "You have been put into the blacklist and may not claim tokens"
+        );
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let claim_start_epoch = self.claim_period_start_epoch().get();
+        let claim_period_in_epochs = self.claim_period_in_epochs().get();
+        let claim_end_epoch = claim_start_epoch + claim_period_in_epochs;
+
+        require!(
+            current_epoch > claim_start_epoch,
+            "Claim period has not started yet"
+        );
+
+        if current_epoch > claim_end_epoch {
```

Comment:
we will not get into claimLaunchpad tokens - until confirming is not finished. might add check for this - like of there are ongoing operations, this endpoint should return error.